### PR TITLE
Improve parse_directory again

### LINKED ISSUE #24: reverse_format has limitations
### Generic Issue

When I write a pattern like "{variable}_{*}_{date_start}.nc", I would like `reverse_format` to fit "pr_gn_2020.nc", but not "pr_v234_gn_2021.nc". Sadly, this is not how it is written and in the second case, `date_start` will be "gn_2021", which breaks everything. 
One solution would be to add as many "{*}" as there are unneeded sections, but that's the point : I don't know. 
My wanted solution would be to forbid the '_' (and '/' and '-') characters in "{*}".

Not sure it would be feasible to fix this in `intake` so maybe there is a better tool somewhere? (like readable regex).

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] `pre-commit` hooks are installed/active in my local clone (`$ pre-commit install`)
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #24
- [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] If a merge request has been made in parallel to this PR in xscen-notebooks, it is merged and the submodules have been updated. (xscen-notebooks#4)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added

### What kind of change does this PR introduce?
Most changes here are to the `parse_directory` mechanism. I made those in parallel to my work on `ouranos_data_catalog`.

* Patterns are matched _relative to_ the path passed in `directories`. If a given file path doesn't have the same "depth" below the `directory` as a pattern, it is skipped. As said on slack, this kinda goes against the idea of the "glob" pattern in folders to skip sub-folders. 
* Wildcards are managed explicitly, which new rules for field matching:
    - The "/" character can never appear in a field.
    - Field starting with "?" or "*" indicate wildcarded fields : they will not be assigned to a column.
    - The "_" character can not appear in a field, except is that field is wildcarded with '*'.
* `cvs_dir` has been changed to `cvs` which is either a yaml file or a dict. Each key must be a column name and they give mapping from parsed vocab to official vocab.
    - In addition, a `attributes` field can be given and it is passed as `attrs_map` to `parse_from_ds`. It provides a mapping from attribute names in the file to column names.
* `_parse_from_zarr` and `_parse_from_nc` added as fast paths for `parse_from_ds`. The first reads the zarr's json metadata directly (which has some limitations) and the second one uses `netCDF4` instead of xarray. Since we do not need to read variables (except `time`), this can be much faster. I have, however, not being able to prove that `_parse_from_nc` is faster than using xarray when `time` is needed, but my tests were very limited.
* `read_from_file` now accepts a list of lists of lists (:roll_eyes:). I.e.: it can read the first file of more than one group definition.
* If a path does not match any pattern, the `name_parser`returns a null Path, and the entry gets dropped.
* If a string date has an unsupported length, `date_parser` returns NaT, instead of failing.
* `globpattern` can try to match folders, in which case the search is two fold: find fitting folders and then find the files. See example in `catalog.ipynb`.
* Update the environment :
    - `h5netcdf` is what xarray needs (h5py is the dependency)
    - Removed the `defaults` channel as it is not needed for xscen.
    - Pinned xESMF
    - Pinned numba 
 
The three last subpoints were done in order to fix the bug where the wrong numpy version gets installed. I am however not sure which one(s) did the trick.

* Added the `only_official_columns` argument to allow non-official fields in the patterns (when False).
   - When `True`, the patterns are checked and all non-wildcarded fields must be in `COLUMNS`, otherwise an error is raised.
   - When `False`, the catalog will have columns based on the fields of the patterns (+ path, format and id).

### Does this PR introduce a breaking change?
The pattern starts where the directory path ends. This makes parsing a subset of directories kinda awkward. See my changes to the notebook : as I now need to parse all CMIP6 data, the `parse_directory` call went from a few seconds to around 25s.

Also, `extension` has been renamed  `globpattern` to better fit its real meaning.

### Other information:
See my upcoming PR on `ouranos_data_catalog` for examples of `parse_directory` calls using some of the new features.

### CODE REVIEW

**Comment by aulemahal** [comment]
Woups j'ai cochonné mes commits un peu... Mais bon l'avant-dernier permet ce dont @RondeauG  avait besoin avec une adaptation de `parse_from_ds`.

File: xscen/catalog.py [hunk 1]
```diff
@@ -679,47 +743,85 @@ def parse_directory(
...
-
+        Parsed directory files
     """
     homogenous_info = homogenous_info or {}
-    columns = set(COLUMNS) - homogenous_info.keys()
-    first_file_only = (
-        None  # The set of columns defining groups for which read the first file.
+    xr_open_kwargs = xr_open_kwargs or {}
+    pattern_fields = set.union(
+        *map(lambda p: set(re.findall(r"{([\w]*)}", p)), patterns)
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/catalog.py:785]: Exact, as * and ? are not part of "\w", which stands for all letters (upper and lower), all numbers and "_".

File: xscen/catalog.py [hunk 2]
```diff
@@ -731,62 +833,70 @@ def delayed_parser(*args, **kwargs):
...
 
-    # translate xrfreq into frequencies and vice-versa
-    df["xrfreq"].fillna(
-        df["frequency"].apply(CV.frequency_to_xrfreq, default=pd.NA), inplace=True
-    )
-    df["frequency"].fillna(
-        df["xrfreq"].apply(CV.xrfreq_to_frequency, default=pd.NA), inplace=True
-    )
+    # Checks
+    # todo
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xscen/catalog.py:898]: Bah. I can see for 1. 
2 is question of accepted vocabulary and 3 may not be a good idea?

File: HISTORY.rst [hunk 3]
```diff
@@ -15,12 +15,14 @@ New features and enhancements
...
-* Add argument ``intermediate_reg_grids`` to ``xscen.regridding.regrid`. (:issue:`34`, :pull:`39`).
-* Add argument ``moving_yearly_window`` to ``xscen.biasadjust.adjust.` (:pull:`39`).
+* Add argument ``intermediate_reg_grids`` to ``xscen.regridding.regrid``. (:issue:`34`, :pull:`39`).
+* Add argument ``moving_yearly_window`` to ``xscen.biasadjust.adjust``. (:pull:`39`).
+* Many adjustments to ``parse_directory``: better wildcards (:issue:`24`), allow custom columns, fastpaths for ``parse_from_ds``, and more (:pull:`30`).
 
 Breaking changes
 ^^^^^^^^^^^^^^^^
-* N/A
+* Patterns in ``parse_directory`` start and the end of the paths in ``directories``. (:pull:`30`).
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply HISTORY.rst:24]: `parallel_depth` has no influence on this!
See this line: https://github.com/Ouranosinc/xscen/blob/4e885ebd7c11629880d966852d07429fc351812e/xscen/catalog.py#L842
And use of `filecoll` elsewhere : the "root" folder is remembered, no matter which subfolders were sent to `find`.

**Review by RondeauG — APPROVED** [review]
It looks good to me!

**Comment by aulemahal** [comment]
Super! Le dernier commit ajoute des balises autodoc/sphinx pour que les constantes (COLUMNS, ID_COLUMNS et autres) soient minimalement documentées dans la doc de l'API.

**Comment by aulemahal** [comment]
Ok j'ai fini.

**Pull request merged**